### PR TITLE
Refactor the way we use RCBI

### PR DESCRIPTION
**Pull request by msmygit**
Refactor the way we use RCBI

### CODE REVIEW

**Comment by absurdfarce** [comment]
So I think we probably _do_ want to move to something like this in the near future.  That idea is really what [this comment](https://github.com/datastax/cassandra-quarkus/pull/229#discussion_r1166087933) was all about.  I'm just reluctant to make this change right before a 3.0.0.Final release; I'd prefer to do it in the _next_ release after, presumably when we've had some time to look into the implications (if any) and better understand what consequences (if any) this change might have.

Make sense?

So I guess I'm saying let's leave this around for now but merge it _after_ 1.2.0 goes out.

File: deployment/src/main/java/com/datastax/oss/quarkus/deployment/internal/CassandraClientProcessor.java [hunk 1]
```diff
@@ -65,24 +65,23 @@ FeatureBuildItem feature() {
...
+                // Should be initialized at build time:
+                "org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerIoRegistryV3d0",
+                "org.apache.tinkerpop.shaded.jackson.databind.deser.std.StdDeserializer",
+                // Required by Tinkerpop:
+                // TODO check if this is really all that is instantiated by reflection
+                "org.apache.tinkerpop.gremlin.structure.Graph",
+                "org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerGraph",
+                "org.apache.tinkerpop.gremlin.structure.util.empty.EmptyGraph",
+                "org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource")
+            .constructors(true)
```
Comment (absurdfarce) [inline deployment/src/main/java/com/datastax/oss/quarkus/deployment/internal/CassandraClientProcessor.java:81]: Looks like constructors(true) was added here when it wasn't present in the original

File: pom.xml [hunk 2]
```diff
@@ -33,7 +33,7 @@
     Note: when upgrading the Quarkus version or the DataStax Java driver version, make sure that you
     upgrade them in the quickstart module too.
     -->
-    <quarkus.version>3.0.0.Final</quarkus.version>
+    <quarkus.version>3.1.0.CR1</quarkus.version>
```
Comment (absurdfarce) [inline pom.xml:36]: Let's revert this change; I'd like to keep this PR focused on the RCBI changes only.

File: quickstart/pom.xml [hunk 3]
```diff
@@ -27,7 +27,7 @@
   <inceptionYear>2020</inceptionYear>
   <properties>
     <java.version>11</java.version>
-    <quarkus.version>3.0.0.Final</quarkus.version>
+    <quarkus.version>3.1.0.CR1</quarkus.version>
```
Comment (absurdfarce) [inline quickstart/pom.xml:30]: Please revert this one as well... same rationale as above.

**Review by absurdfarce — CHANGES_REQUESTED** [review]
Looks pretty sane over all, just a couple of minor changes and I think we'll be all set.

File: deployment/src/main/java/com/datastax/oss/quarkus/deployment/internal/CassandraClientProcessor.java (hunk 1 above)
Comment (msmygit) [reply deployment/src/main/java/com/datastax/oss/quarkus/deployment/internal/CassandraClientProcessor.java:81]: Because previously constructors were true by default and that [method is now deprecated](https://github.com/quarkusio/quarkus/blob/3.0.3.Final/core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ReflectiveClassBuildItem.java#L80-L87) and hence the change.

Comment (msmygit) [reply deployment/src/main/java/com/datastax/oss/quarkus/deployment/internal/CassandraClientProcessor.java:81]: Before you ask, even in [2.x Quarkus version, constructors were always `true`](https://github.com/quarkusio/quarkus/blob/2.16.7.Final/core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ReflectiveClassBuildItem.java#L24-L26).

Comment (absurdfarce) [reply deployment/src/main/java/com/datastax/oss/quarkus/deployment/internal/CassandraClientProcessor.java:81]: The specific constructor you cite for 3.0 does calls another constructor in turn and it does set the "constructors" param to true when it does so.  But that's only because the constructor in question doesn't accept a "constructors" argument.  [Other](https://github.com/quarkusio/quarkus/blob/3.0.3.Final/core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ReflectiveClassBuildItem.java#L71-L78) [constructors](https://github.com/quarkusio/quarkus/blob/3.0.3.Final/core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ReflectiveClassBuildItem.java#L89-L96) do accept such an argument.

It's true that these methods are deprecated, but this appears to be the case for all RCBI constructors; based on the comments it's clear they're trying to steer developers to use the builders instead (which is entirely sensible).  The builder still very much includes a "constructors" [method](https://github.com/quarkusio/quarkus/blob/3.0.3.Final/core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ReflectiveClassBuildItem.java#L196-L199) which accepts a boolean argument but [the default](https://github.com/quarkusio/quarkus/blob/3.0.3.Final/core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ReflectiveClassBuildItem.java#L181) is still "true".

So at best including "constructors(true)" here is redundant (since it's just restating the default).

With respect to Quarkus 2.x: the set of constructors there look very similar (they just aren't deprecated yet).  As discussed above some of the constructors set "constructors" to true if they don't accept that as an argument but there are plenty of constructors in 2.x which do accept a boolean arg corresponding to "constructors"... so it very much comes down to which constructor a developer wants to make use of.

I guess I'd argue against changing the builder methods called here unless we have a clear benefit from it.  Seem sensible?

Comment (msmygit) [reply deployment/src/main/java/com/datastax/oss/quarkus/deployment/internal/CassandraClientProcessor.java]: Good point. Removed calling the `.constructors(true)` from here.

**Comment by absurdfarce** [comment]
I just noticed: we need to change this PR around to go against main.  quarkus-300Final has been [merged into main](https://github.com/datastax/cassandra-quarkus/commit/43ad2c06bf44a4d1bb0e5448db75b88b2688d828) already.  In fact we probably ought to just delete the quarkus-300Final branch outright.

**Comment by msmygit** [comment]
Closing this in favor of #236
